From cef093f9b849c65b231bd6f29eda3bdefedfabb1 Mon Sep 17 00:00:00 2001
From: Richard Zhu <Richard.Zhu@freescale.com>
Date: Tue, 18 Aug 2015 08:34:15 +0800
Subject: [PATCH 0410/5242] MLK-11498 rpmsg: imx: tell remote process about
 the channel

commit  1580c375a7d43e6c01e8d9538e883cfdd50ac129 from
https://source.codeaurora.org/external/imx/linux-imx.git

send a message to our remote processor, and tell remote
processor about this channel

Signed-off-by: Richard Zhu <Richard.Zhu@freescale.com>
(cherry picked from commit 2708c004a60c5b6da020803ee9291b83984d4a65)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/rpmsg/Kconfig              |    4 ++--
 drivers/rpmsg/imx_rpmsg_pingpong.c |   15 +++++++++++++--
 drivers/rpmsg/imx_rpmsg_tty.c      |   11 +++++++++++
 3 files changed, 26 insertions(+), 4 deletions(-)

diff --git a/drivers/rpmsg/Kconfig b/drivers/rpmsg/Kconfig
index f314c69..7352af7 100644
--- a/drivers/rpmsg/Kconfig
+++ b/drivers/rpmsg/Kconfig
@@ -56,11 +56,11 @@ config RPMSG_VIRTIO
 	select VIRTIO
 
 config IMX_RPMSG_PINGPONG
-	tristate "IMX RPMSG pingpong driver"
+	tristate "IMX RPMSG pingpong driver -- loadable modules only"
 	depends on RPMSG && m
 
 config IMX_RPMSG_TTY
-	tristate "IMX RPMSG tty driver"
+	tristate "IMX RPMSG tty driver -- loadable modules only"
 	depends on RPMSG && m
 
 endmenu
diff --git a/drivers/rpmsg/imx_rpmsg_pingpong.c b/drivers/rpmsg/imx_rpmsg_pingpong.c
index 766de0b..341cd7e 100644
--- a/drivers/rpmsg/imx_rpmsg_pingpong.c
+++ b/drivers/rpmsg/imx_rpmsg_pingpong.c
@@ -17,6 +17,7 @@
 #include <linux/virtio.h>
 #include <linux/rpmsg.h>
 
+#define MSG		"hello world!"
 #define MSG_LIMIT	100000
 static unsigned int rpmsg_pingpong;
 static int rx_count;
@@ -41,7 +42,7 @@ static int rpmsg_pingpong_cb(struct rpmsg_device *rpdev, void *data, int len,
 	err = rpmsg_sendto(rpdev->ept, (void *)(&rpmsg_pingpong), 4, src);
 
 	if (err)
-		pr_err("rpmsg_send failed: %d\n", err);
+		dev_err(&rpdev->dev, "rpmsg_send failed: %d\n", err);
 
 	return err;
 }
@@ -53,11 +54,21 @@ static int rpmsg_pingpong_probe(struct rpmsg_device *rpdev)
 	dev_info(&rpdev->dev, "new channel: 0x%x -> 0x%x!\n",
 			rpdev->src, rpdev->dst);
 
+	/*
+	 * send a message to our remote processor, and tell remote
+	 * processor about this channel
+	 */
+	err = rpmsg_send(rpdev->ept, MSG, strlen(MSG));
+	if (err) {
+		dev_err(&rpdev->dev, "rpmsg_send failed: %d\n", err);
+		return err;
+	}
+
 	rpmsg_pingpong = 0;
 	rx_count = 0;
 	err = rpmsg_sendto(rpdev->ept, (void *)(&rpmsg_pingpong), 4, rpdev->dst);
 	if (err) {
-		pr_err("rpmsg_send failed: %d\n", err);
+		dev_err(&rpdev->dev, "rpmsg_send failed: %d\n", err);
 		return err;
 	}
 
diff --git a/drivers/rpmsg/imx_rpmsg_tty.c b/drivers/rpmsg/imx_rpmsg_tty.c
index 26d3b14..658be55 100644
--- a/drivers/rpmsg/imx_rpmsg_tty.c
+++ b/drivers/rpmsg/imx_rpmsg_tty.c
@@ -35,6 +35,7 @@ struct rpmsgtty_port {
 
 /* this needs to be less then (RPMSG_BUF_SIZE - sizeof(struct rpmsg_hdr)) */
 #define RPMSG_MAX_SIZE		256
+#define MSG		"hello world!"
 
 static int rpmsg_tty_cb(struct rpmsg_device *rpdev, void *data, int len,
 						void *priv, u32 src)
@@ -144,6 +145,16 @@ static int rpmsg_tty_probe(struct rpmsg_device *rpdev)
 	dev_info(&rpdev->dev, "new channel: 0x%x -> 0x%x!\n",
 			rpdev->src, rpdev->dst);
 
+	/*
+	 * send a message to our remote processor, and tell remote
+	 * processor about this channel
+	 */
+	err = rpmsg_send(rpdev->ept, MSG, strlen(MSG));
+	if (err) {
+		dev_err(&rpdev->dev, "rpmsg_send failed: %d\n", err);
+		return err;
+	}
+
 	rpmsgtty_driver = tty_alloc_driver(1, TTY_DRIVER_UNNUMBERED_NODE);
 	if (IS_ERR(rpmsgtty_driver))
 		return PTR_ERR(rpmsgtty_driver);
-- 
1.7.9.5

